external-crates/move/crates/move-package/src/resolution/resolution_graph.rs

```diff
@@ -345,19 +349,52 @@ impl Package {
         })
     }
 
-    fn define_addresses_in_package(&self, resolving_table: &mut ResolvingTable) -> Result<()> {
+    /// Associates addresses with named packages in the `resolving_table`.
+    /// Addresses may be pulled in from two sources:
+    /// - The [addresses] section `Move.toml`.
+    /// - Adresses (package IDs) in the `Move.lock` associated with published packages for `chain_id`.
+    ///
+    /// Addresses are pulled from the `Move.lock` only when a package is published or upgraded on-chain.
+    /// Local builds only consult the `Move.toml` manifest.
+    fn define_addresses_in_package(
+        &self,
+        resolving_table: &mut ResolvingTable,
+        chain_id: &Option<String>,
+    ) -> Result<()> {
         let pkg_id = custom_resolve_pkg_id(&self.source_package).with_context(|| {
             format!(
                 "Resolving package name for '{}'",
                 &self.source_package.package.name
             )
         })?;
         for (name, addr) in self.source_package.addresses.iter().flatten() {
+            if *addr == Some(AccountAddress::ZERO) {
+                // The address in the manifest is set to 0x0, meaning `name` is associated with 'this'
+                // package. Published dependent package IDs are resolved by `chain_id` from the
+                // `Move.lock` when a package is to be published or upgraded.
+                if let Some(original_id) = self.resolve_original_id_from_lock(chain_id) {
+                    let addr = AccountAddress::from_str(&original_id)?;
+                    resolving_table.define((pkg_id, *name), Some(addr))?;
+                    continue;
+                }
+            }
```

Comment:
This is Part 1: identify when we should resolve original IDs from the `Move.lock`, and do so when possible for the respective `chain_id`.

Nitty gritty note: the first commit in this PR propagates `chain_id` to this part of the logic: ff73dbfbc8e759c8c3ffc44a7299280f34d6d12a. It might work to pass this value in `BuildConfig` instead of propagating through function args. The change for propagating through functions didn't feel too invasive though, so went with that.